// sudoku-master/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VALID_DIGITS = [1, 2, 3, 4, 5, 6, 7, 8, 9];
exports.VALID_LINE_TYPES = ["row", "col"];
exports.VALID_HOUSE_TYPES = [...exports.VALID_LINE_TYPES, "box"];
exports.VALID_HOUSE_INDEXES = [0, 1, 2, 3, 4, 5, 6, 7, 8];
exports.VALID_CELL_INDEXES = [...exports.VALID_HOUSE_INDEXES];
exports.ROW1_INDEXES = [0, 1, 2, 3, 4, 5, 6, 7, 8];
exports.ROW2_INDEXES = [9, 10, 11, 12, 13, 14, 15, 16, 17];
exports.ROW3_INDEXES = [18, 19, 20, 21, 22, 23, 24, 25, 26];
exports.ROW4_INDEXES = [27, 28, 29, 30, 31, 32, 33, 34, 35];
exports.ROW5_INDEXES = [36, 37, 38, 39, 40, 41, 42, 43, 44];
exports.ROW6_INDEXES = [45, 46, 47, 48, 49, 50, 51, 52, 53];
exports.ROW7_INDEXES = [54, 55, 56, 57, 58, 59, 60, 61, 62];
exports.ROW8_INDEXES = [63, 64, 65, 66, 67, 68, 69, 70, 71];
exports.ROW9_INDEXES = [72, 73, 74, 75, 76, 77, 78, 79, 80];
exports.VALID_GRID_INDEXES = [
    ...exports.ROW1_INDEXES,
    ...exports.ROW2_INDEXES,
    ...exports.ROW3_INDEXES,
    ...exports.ROW4_INDEXES,
    ...exports.ROW5_INDEXES,
    ...exports.ROW6_INDEXES,
    ...exports.ROW7_INDEXES,
    ...exports.ROW8_INDEXES,
    ...exports.ROW9_INDEXES,
];
