// sudoku-master/lib/utils/house/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const remeda_1 = require("remeda");
const types_1 = require("../../types");
const getCellIndexInGrid_1 = require("../cell/getCellIndexInGrid");
exports.HOUSES_LIST = remeda_1.flatMap((houseType) => remeda_1.map((houseIndex) => {
    return {
        type: houseType,
        index: houseIndex,
        cells: remeda_1.map(types_1.VALID_CELL_INDEXES, (cellIndex) => {
            const indexInGrid = getCellIndexInGrid_1.getCellIndexInGrid({ houseType, houseIndex, cellIndex });
            return indexInGrid;
        }),
    };
})(types_1.VALID_HOUSE_INDEXES))(types_1.VALID_HOUSE_TYPES);
exports.ROWS_LIST = remeda_1.pipe(exports.HOUSES_LIST, remeda_1.filter((house) => house.type === "row"), remeda_1.sortBy((house) => house.index));
exports.COLUMNS_LIST = remeda_1.pipe(exports.HOUSES_LIST, remeda_1.filter((house) => house.type === "col"), remeda_1.sortBy((house) => house.index));
exports.BOXES_LIST = remeda_1.pipe(exports.HOUSES_LIST, remeda_1.filter((house) => house.type === "box"), remeda_1.sortBy((house) => house.index));
